state/upgrades.go

```diff
@@ -253,3 +254,45 @@ func EnsureInitalRefCountForExternalSecretBackends(pool *StatePool) error {
 	}
 	return nil
 }
+
+func EnsureApplicationCharmOriginsHaveRevisions(pool *StatePool) error {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		allApps, err := st.AllApplications()
+		if err != nil {
+			return errors.Annotate(err, "loading applications")
+		}
+
+		var ops []txn.Op
+
+		for _, app := range allApps {
+			// We only need to fill in the revision if it's nil/
+			// There is an edge case though where a previous migration
+			// has incorrectly filled in the revision to 0 or -1
+			rev := app.CharmOrigin().Revision
+			if rev != nil && *rev > 0 {
+				continue
+			}
+			curlStr, _ := app.CharmURL()
+			if curlStr == nil {
+				return errors.Errorf("application %q has no charm url", app.Name())
+			}
+			curl, err := charm.ParseURL(*curlStr)
+			if err != nil {
+				return errors.Annotatef(err, "parsing charm url %q", *curlStr)
+			}
+			if curl.Revision == -1 {
+				return errors.Errorf("charm url %q has no revision", curl.String())
+			}
+			ops = append(ops, txn.Op{
+				C:      applicationsC,
+				Id:     app.doc.DocID,
+				Assert: bson.D{{"charm-origin.revision", bson.D{{"$exists", false}}}},
```

Comment:
This check is most likely causing the failure.

Reply:
You're right, originally it was valid, but we also need to check & fill in revisions equal to 0